edb/server/protocol/request_scheduler.py

```diff
@@ -0,0 +1,570 @@
+
+        if self.request_limits is not None:
+            base_delay = self.request_limits.base_delay(
+                deferred_cost, guess=self.guess_delay,
+            )
+            if base_delay is None:
+                delay = None
+
+            else:
+                # If delay_factor is very high, it may take quite a long time
+                # for it to return to 1. A maximum delay prevents this service
+                # from never getting checked.
+                delay = min(
+                    base_delay * self.request_limits.delay_factor,
+                    self.delay_max,
+                )
+
+        else:
+            # We have absolutely no information about the delay, assume naptime.
+            delay = naptime
+
+        if error_count > 0:
+            # There was an error, wait before trying again.
+            # Use the larger of delay or naptime.
+            delay = max(delay, naptime) if delay is not None else naptime
+            urgent = False
+
+        elif deferred_cost > 0:
+            # There is some deferred work, apply the delay and run immediately.
+            urgent = True
+
+        elif success_count > 0:
+            # Some work was done successfully. Run again to ensure no more work
+            # needs to be done.
+            delay = None
+            urgent = True
+
+        else:
+            # No work left to do. Take a nap.
+            delay = naptime
+            urgent = False
+
+        return Timer.create_delay(delay, urgent)
+
+
+@dataclass
+class Limits:
+    """Information about a service's rate limits."""
+
+    # Total limit of a resource per minute for a service.
+    # A True value represents a unlimited total
+    total: Optional[int | Literal[True]] = None
+
+    # Remaining resources before the limit is hit.
+    remaining: Optional[int] = None
+
+    # A delay factor to implement exponential backoff
+    delay_factor: float = 1
+
+    def base_delay(
+        self,
+        request_cost: int,
+        *,
+        guess: float,
+    ) -> Optional[float]:
+        if self.total is True:
+            return None
+
+        if self.remaining is not None and request_cost <= self.remaining:
+            return None
+
+        if self.total is not None:
+            return 60.0 / self.total * 1.1  # 10% buffer just in case
+
+        # guess the delay
+        return guess
+
+    def update(self, latest: Limits) -> Limits:
+        """Update based on the latest information."""
+
+        # The total will change rarely. Always take the latest value if
+        # it exists.
+        if latest.total is not None:
+            self.total = latest.total
+
+        # The remaining amount can fluctuate quite a bit, take the smallest
+        # value available.
+        if self.remaining is None:
+            self.remaining = latest.remaining
+        elif latest.remaining is not None:
+            self.remaining = min(self.remaining, latest.remaining)
```

Comment:
The logic around updating `remaining` is a bit strange.

Here are the goals:
- Optionally set a remaining amount before calling the schedulers,
- Update or set the remaining amount on the fly if the service returns one
- Clear the remaining amount after executing the requests

I am assuming that `execute_no_sleep` is fast and that the rate limit is low enough that the remaining amount will only decrease during the call, even with retries.

I suppose this could be a bit inefficient if a service has a very very high, but never provides a remaining limit. In that case, only a single request will be handled before `Scheduler.process` needs to be called again. `Service.next_delay` will report that it needs to be called again right away, so that should be ok I think.

Reply:
Please add some extra discussion here in a comment. I think the key thing I missed is that `remaining` is a totally transient thing: used during submission of stuff but not tracked across sleeps